File: tests/unit/src/android/telephony/imsmedia/AnbrModeTest.java
/*
 * Copyright (C) 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.telephony.imsmedia;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNotNull;

import android.os.Parcel;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class AnbrModeTest {

    private static final int ANBR_UPLINK_MODE = 1;
    private static final int ANBR_DOWNLINK_MODE = 2;

    @Test
    public void testBuilder() {
        AnbrMode anbrMode = new AnbrMode.Builder()
                .setAnbrUplinkCodecMode(ANBR_UPLINK_MODE)
                .setAnbrDownlinkCodecMode(ANBR_DOWNLINK_MODE)
                .build();

        assertEquals(ANBR_UPLINK_MODE, anbrMode.getAnbrUplinkCodecMode());
        assertEquals(ANBR_DOWNLINK_MODE, anbrMode.getAnbrDownlinkCodecMode());
    }

    @Test
    public void testSettersAndGetters() {
        AnbrMode anbrMode = new AnbrMode.Builder().build();

        anbrMode.setAnbrUplinkCodecMode(ANBR_UPLINK_MODE);
        assertEquals(ANBR_UPLINK_MODE, anbrMode.getAnbrUplinkCodecMode());

        anbrMode.setAnbrDownlinkCodecMode(ANBR_DOWNLINK_MODE);
        assertEquals(ANBR_DOWNLINK_MODE, anbrMode.getAnbrDownlinkCodecMode());
    }

    @Test
    public void testParcelable() {
        AnbrMode anbrMode = new AnbrMode.Builder()
                .setAnbrUplinkCodecMode(ANBR_UPLINK_MODE)
                .setAnbrDownlinkCodecMode(ANBR_DOWNLINK_MODE)
                .build();

        Parcel parcel = Parcel.obtain();
        anbrMode.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);

        AnbrMode fromParcel = AnbrMode.CREATOR.createFromParcel(parcel);

        assertEquals(anbrMode, fromParcel);
        parcel.recycle();
    }

    @Test
    public void testEqualsAndHashCode() {
        AnbrMode anbrMode1 = new AnbrMode.Builder()
                .setAnbrUplinkCodecMode(ANBR_UPLINK_MODE)
                .setAnbrDownlinkCodecMode(ANBR_DOWNLINK_MODE)
                .build();

        AnbrMode anbrMode2 = new AnbrMode.Builder()
                .setAnbrUplinkCodecMode(ANBR_UPLINK_MODE)
                .setAnbrDownlinkCodecMode(ANBR_DOWNLINK_MODE)
                .build();

        AnbrMode anbrMode3 = new AnbrMode.Builder()
                .setAnbrUplinkCodecMode(ANBR_UPLINK_MODE + 1)
                .setAnbrDownlinkCodecMode(ANBR_DOWNLINK_MODE)
                .build();

        AnbrMode anbrMode4 = new AnbrMode.Builder()
                .setAnbrUplinkCodecMode(ANBR_UPLINK_MODE)
                .setAnbrDownlinkCodecMode(ANBR_DOWNLINK_MODE + 1)
                .build();

        assertEquals(anbrMode1, anbrMode2);
        assertEquals(anbrMode1.hashCode(), anbrMode2.hashCode());

        assertNotEquals(anbrMode1, anbrMode3);
        assertNotEquals(anbrMode1.hashCode(), anbrMode3.hashCode());

        assertNotEquals(anbrMode2, anbrMode4);
        assertNotEquals(anbrMode1.hashCode(), anbrMode4.hashCode());

        assertNotEquals(anbrMode3, anbrMode4);
        assertNotEquals(anbrMode3.hashCode(), anbrMode4.hashCode());
    }

    @SuppressWarnings("SelfAssertion")
    @Test
    public void testSpecialEqualsCases() {
        AnbrMode anbrMode = new AnbrMode.Builder()
                .setAnbrUplinkCodecMode(ANBR_UPLINK_MODE)
                .setAnbrDownlinkCodecMode(ANBR_DOWNLINK_MODE)
                .build();
        // Test equals with the same object reference
        assertEquals(anbrMode, anbrMode);
        // Test equals with null
        assertNotEquals(anbrMode, null);
        // Test equals with a different object type
        assertNotEquals(anbrMode, new Object());
    }

    @Test
    public void testToString() {
        AnbrMode anbrMode = new AnbrMode.Builder()
                .setAnbrUplinkCodecMode(ANBR_UPLINK_MODE)
                .setAnbrDownlinkCodecMode(ANBR_DOWNLINK_MODE)
                .build();

        String expected = "AnbrMode: {mAnbrUplinkMode =" + ANBR_UPLINK_MODE
                + ", mAnbrDownlinkMode =" + ANBR_DOWNLINK_MODE + " }";
        assertEquals(expected, anbrMode.toString());
    }

    @Test
    public void testDescribeContents() {
        AnbrMode anbrMode = new AnbrMode.Builder().build();
        assertEquals(0, anbrMode.describeContents());
    }

    @Test
    public void testNewArray() {
        final int size = 5;
        AnbrMode[] anbrModes = AnbrMode.CREATOR.newArray(size);
        assertNotNull(anbrModes);
        assertEquals(size, anbrModes.length);
    }
}

File: tests/native/service/src/com/android/telephony/imsmedia/lib/libimsmedia/config/AnbrModeTest.cpp
/**
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <AnbrMode.h>
#include <gtest/gtest.h>

using namespace android::telephony::imsmedia;
using namespace android;

const int32_t kAnbrUplinkMode = 1;
const int32_t kAnbrDownlinkMode = 2;

TEST(AnbrModeTest, TestConstructor)
{
    AnbrMode mode;
    EXPECT_EQ(mode.getAnbrUplinkCodecMode(), 0);
    EXPECT_EQ(mode.getAnbrDownlinkCodecMode(), 0);

    AnbrMode* mode2 = new AnbrMode(mode);
    EXPECT_EQ(mode, *mode2);
    delete mode2;
}

TEST(AnbrModeTest, TestDefault)
{
    AnbrMode mode;
    mode.setDefaultAnbrMode();
    EXPECT_EQ(mode.getAnbrUplinkCodecMode(), 0);
    EXPECT_EQ(mode.getAnbrDownlinkCodecMode(), 0);
}

TEST(AnbrModeTest, TestGetterSetter)
{
    AnbrMode mode;
    mode.setAnbrUplinkCodecMode(kAnbrUplinkMode);
    mode.setAnbrDownlinkCodecMode(kAnbrDownlinkMode);
    EXPECT_EQ(mode.getAnbrUplinkCodecMode(), kAnbrUplinkMode);
    EXPECT_EQ(mode.getAnbrDownlinkCodecMode(), kAnbrDownlinkMode);
}

TEST(AnbrModeTest, TestParcel)
{
    AnbrMode mode1;
    mode1.setAnbrUplinkCodecMode(kAnbrUplinkMode);
    mode1.setAnbrDownlinkCodecMode(kAnbrDownlinkMode);

    android::Parcel parcel;
    EXPECT_EQ(mode1.writeToParcel(nullptr), BAD_VALUE);
    EXPECT_EQ(mode1.writeToParcel(&parcel), NO_ERROR);
    parcel.setDataPosition(0);

    AnbrMode mode2;
    EXPECT_EQ(mode2.readFromParcel(nullptr), BAD_VALUE);
    EXPECT_EQ(mode2.readFromParcel(&parcel), NO_ERROR);
    EXPECT_EQ(mode1, mode2);
}

TEST(AnbrModeTest, TestAssign)
{
    AnbrMode mode1;
    mode1.setAnbrUplinkCodecMode(kAnbrUplinkMode);
    mode1.setAnbrDownlinkCodecMode(kAnbrDownlinkMode);

    AnbrMode mode2;
    mode2 = mode1;
    EXPECT_EQ(mode1, mode2);
}

TEST(AnbrModeTest, TestEqual)
{
    AnbrMode mode1;
    mode1.setAnbrUplinkCodecMode(kAnbrUplinkMode);
    mode1.setAnbrDownlinkCodecMode(kAnbrDownlinkMode);

    AnbrMode mode2;
    mode2.setAnbrUplinkCodecMode(kAnbrUplinkMode);
    mode2.setAnbrDownlinkCodecMode(kAnbrDownlinkMode);
    EXPECT_EQ(mode1, mode2);
}

TEST(AnbrModeTest, TestNotEqual)
{
    AnbrMode mode1;
    mode1.setAnbrUplinkCodecMode(kAnbrUplinkMode);
    mode1.setAnbrDownlinkCodecMode(kAnbrDownlinkMode);

    AnbrMode mode2;
    mode2.setAnbrUplinkCodecMode(3);
    mode2.setAnbrDownlinkCodecMode(kAnbrDownlinkMode);

    AnbrMode mode3;
    mode3.setAnbrUplinkCodecMode(kAnbrUplinkMode);
    mode3.setAnbrDownlinkCodecMode(4);

    EXPECT_NE(mode1, mode2);
    EXPECT_NE(mode1, mode3);
}
